### PR TITLE
Models

### PR DESCRIPTION
**Pull request by uzair-imtiaz**
### Description
This pull request introduces several models that enhance profile management and project feedback features to the application. The new models include Profile, Skill, Project, Review, and Vote, which collectively contribute to the main functionality of the app, i.e, create user profiles, upload projects, leave your votes and reviews on the projects.

### Testing Instructions
To test the newly added features and ensure their proper functioning, follow the steps below:

Clone the repository:
`git clone git@github.com:Uzair-Imtiaz/Code-Book.git`

Create a virtual environment:
`python3 -m venv myenv`

Activate the venv:
`source myenv/bin/activate`

Install the required modules:
`pip install -r requirements.txt`

Navigate to the project directory:
`cd Code-Book`

Checkout to the models branch:
`git checkout models`

Create a super user so that you can have access to the admin interface of the database.
`python3 manage.py createsuperuser`

    - Provide a username.
    - Email is optional.
    - Provide a password.
    - Confirm password.
 Success message will be displayed if the user is created.

Apply database migrations:
`python3 manage.py migrate`

Start the development server:
`python3 manage.py runserver`

Open a web browser and visit the application at **http://localhost:8000/**.

To see the models, visit **http://localhost:8000/admin** and login with the superuser credentials you created.

### Snapshots
Screenshot 2023-07-09 at 2 48 22 PM

### CODE REVIEW

File: authentication/models.py [hunk 1]
```diff
@@ -1,5 +1,106 @@
...
+
+    Fields:
+       - user: The related user.
+       - skills: The list of skills associated with the profile.
+       - bio: The biography of the user.
+       - profile_picture : The profile picture of the user.
+       - github: The GitHub profile link of the user.
+       - linkedin: The LinkedIn profile link of the user.
+       - youtube: The YouTube profile link of the user.
+       - gender: The gender of the user.
```
Comment (Ali-D-Akbar) [inline authentication/models.py]: This can be removed as we already have help text.

File: authentication/models.py [hunk 2]
```diff
@@ -1,5 +1,106 @@
...
+        'Skill',
+        blank=True,
+        null=True,
+        related_name='profile',
+        help_text='The skills associated with the profile.'
+    )
+    bio = models.TextField(
+        blank=True,
+        null=True,
+        help_text='A detailed bio of the user about himself.'
```
Comment (Ali-D-Akbar) [inline authentication/models.py]: improve the help text to be generic rather than gender specific.

File: authentication/models.py [hunk 3]
```diff
@@ -1,5 +1,106 @@
...
+    )
+    profile_picture = models.ImageField(
+        blank=True,
+        null=True,
+        help_text='A profile picture of the user.'
+    )
+    github = models.URLField(
```
Comment (Ali-D-Akbar) [inline authentication/models.py:35]: Github username can also be used.

File: authentication/models.py [hunk 4]
```diff
@@ -1,5 +1,106 @@
...
+        blank=True,
+        null=True,
+        help_text='The date of birth of the user.'
+    )
+
+    def __str__(self):
+        return f"{self.user.first_name} {self.user.last_name}"
+
+
+class Skill(models.Model):
```
Comment (Ali-D-Akbar) [inline authentication/models.py]: Use Timestamped model for metadata dates like created/updated.

File: authentication/models.py [hunk 5]
```diff
@@ -1,5 +1,106 @@
-""" Models for the authentication app."""
+""" This module contains models for the authentication app. """
 
 from django.db import models
 from django.contrib.auth.models import User
-# Create your models here.
+
+
+class Profile(models.Model):
```
Comment (Ali-D-Akbar) [inline authentication/models.py]: Use Timestamped Model

File: authentication/views.py [hunk 6]
```diff
@@ -1,9 +1,20 @@
...
+Views for user registration, login, and welcome in the 'authentication' app.
 
+This file contains view classes and a function for handling user registration, login, and rendering the welcome page.
+
+Classes:
+- RegisterView(View): View class for user registration.
+- LoginView(View): View class for user login.
+
+Functions:
+- welcome(request): Renders the welcome page.
```
Comment (Ali-D-Akbar) [inline authentication/views.py]: Can be omitted.

File: core/models.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+    Fields:
+    - user: The user that created the project.
+    - title: The title of the project.
+    - description: The description of the project.
+    - featured_image: The featured image of the project.
+    - youtube_link: The link to the YouTube video of the project.
+    - demo_link: The link to the demo of the project.
+    - source_code_link: The link to the source code of the project.
+    """
```
Comment (Ali-D-Akbar) [inline core/models.py]: Can be omitted.

**Review by Ali-D-Akbar — COMMENTED** [review]
A bunch of changes before final approval. Yet to verify on local.

File: authentication/models.py (hunk 3 above)
Comment (uzair-imtiaz) [reply authentication/models.py:35]: I was thinking to use these fields in the profile cards as href that's why I took them as URLField.

File: core/migrations/0001_initial.py [hunk 8]
```diff
@@ -1,8 +1,8 @@
 # Generated by Django 4.2.2 on 2023-07-09 09:43
 
+import django.db.models.deletion
```
Comment (Ali-D-Akbar) [inline core/migrations/0001_initial.py:3]: Migrations ideally shouldn't change so you can exclude them from the quality checks.

**Review by Ali-D-Akbar — APPROVED** [review]
One comment otherwise, LGTM.

**Pull request merged**